zkevm-circuits/src/evm_circuit/execution/return.rs

```diff
@@ -24,25 +49,342 @@ impl<F: Field> ExecutionGadget<F> for ReturnGadget<F> {
+
+        // These are globally defined because they are used across multiple cases.
+        let copy_rw_increase = cb.query_cell();
+        let copy_rw_increase_is_zero = IsZeroGadget::construct(cb, copy_rw_increase.expr());
 
-        Self { opcode }
+        // Case A in the specs.
+        cb.condition(is_create.expr() * is_success.expr(), |cb| {
+            cb.require_equal(
+                "increase rw counter once for each memory to bytecode byte copied",
+                copy_rw_increase.expr(),
+                range.length(),
+            );
+        });
+        cb.condition(
+            is_create.expr() * is_success.expr() * not::expr(copy_rw_increase_is_zero.expr()),
+            |_cb| {
+                // TODO: copy_table_lookup for contract creation.
+            },
+        );
+
+        // Case B in the specs.
+        cb.condition(is_root.expr(), |cb| {
+            cb.require_next_state(ExecutionState::EndTx);
+        });
+
+        // Case C in the specs.
+        // TODO: have copy_table_lookup update rw_counter expression so that this can go
+        // at the end of the constraints.
+        let restore_context = cb.condition(not::expr(is_root.expr()), |cb| {
+            RestoreContextGadget::construct(
+                cb,
+                is_success.expr(),
+                3.expr() + copy_rw_increase.expr(),
+                range.offset(),
+                range.length(),
+            )
+        });
+
+        // Case D in the specs.
+        let (caller_id, return_data_offset, return_data_length, copy_length) = cb.condition(
+            not::expr(is_create.expr()) * not::expr(is_root.expr()),
+            |cb| {
+                let [caller_id, return_data_offset, return_data_length] = [
+                    CallContextFieldTag::CallerId,
+                    CallContextFieldTag::ReturnDataOffset,
+                    CallContextFieldTag::ReturnDataLength,
+                ]
+                .map(|field_tag| cb.call_context(None, field_tag));
+                let copy_length =
+                    MinMaxGadget::construct(cb, return_data_length.expr(), range.length());
+                cb.require_equal(
+                    "increase rw counter twice for each memory to memory byte copied",
+                    copy_length.min() + copy_length.min(),
+                    copy_rw_increase.expr(),
+                );
+                (
+                    caller_id,
+                    return_data_offset,
+                    return_data_length,
+                    copy_length,
+                )
+            },
+        );
+        cb.condition(
+            not::expr(is_create.expr())
+                * not::expr(is_root.expr())
+                * not::expr(copy_rw_increase_is_zero.expr()),
+            |cb| {
+                cb.copy_table_lookup(
+                    cb.curr.state.call_id.expr(),
+                    CopyDataType::Memory.expr(),
+                    caller_id.expr(),
```

Comment:
Couldn't you skip a lookup and use `cb.next.state.call_id.expr()` here, like in the spec?
https://github.com/privacy-scaling-explorations/zkevm-specs/blob/677305ecb845cfec8b91cd52195af576feb29564/src/zkevm_specs/evm/execution/return_.py#L68

Reply:
I wasn't aware that was available on the constraint builder :)